Add 'service' subcommand with associated operations

miguelsorianod opened the pull request:
Adds the 'service' subcommand in the CLI.

This should cover issue https://github.com/3scale/3scale_toolbox/issues/40

* Pending to add tests

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,33 @@
+module ThreeScaleToolbox
+  module Commands
+    module ServiceCommand
+      class ServiceCreateSubcommand < Cri::CommandRunner
```
In the old CLI the 'create' command had the following flags:

```
    -s, --service <service_id>        Specify service id // Although seems to not be used in the command even if you specify it
    -c, --serviceName <service_name>  Specify service name
    -a, --authentication <mode>       Specify authentication mode ('1' for API key, '2' for App Id / App Key, 'oauth' for OAuth mode, 'oidc' for OpenID Connect)
```

I changed the -c flag for a positional parameter because it is mandatory to specify at least the service name.
Is the 'authentication' flag used nowadays? should we add it?

eguzki replied:
I would allow as optional params: 
* system name (default `service_name.downcase.gsub(/[^\w]/, '_')`)
* authentication: `backend_version` attr. (default: '1' for apiKey). 
* gateway: `deployment_mode` attr. (Default: 'hosted')

miguelsorianod replied:
:+1:

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,33 @@
+            runner ServiceCreateSubcommand
+
+            param   :source
+            param   :service_name
```
The command format is:
`3scale service create  [flags] <source> <service-name>`

eguzki replied:
:+1:

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_list.rb:
```diff
@@ -0,0 +1,32 @@
+            description 'List all services'
+            runner ServiceListSubcommand
+
+            param   :source
```
Command format is:
`3scale service list [flags] <source>`

eguzki replied:
:+1:

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_remove.rb:
```diff
@@ -0,0 +1,32 @@
+            runner ServiceRemoveSubcommand
+
+            param   :source
+            param   :service_id
```
Format of the command:

`3scale service remove <source> <service-id>`

eguzki replied:
`<service-id_or_system-name>`

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_show.rb:
```diff
@@ -0,0 +1,53 @@
+          end
+        end
+
+        SERVICE_FIELDS_TO_SHOW = %w[
```
I've seen a very similar array was defined in https://github.com/3scale/3scale_toolbox/blob/7a5e540e2919bd069ba416f27139daa654211f47/lib/3scale_toolbox/entities/service.rb#L4
although that is used for another purposes and some of the fields are omitted.
Do you think we should move that variable there? split arrays or something?
It seems very specific of the specific output of the command although they are field names contained in the result.

eguzki replied:
Maybe in `ThreeScaleToolbox::Commands::ServiceCommand` namespace??

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_update.rb:
```diff
@@ -0,0 +1,28 @@
+module ThreeScaleToolbox
+  module Commands
+    module ServiceCommand
+      class ServiceUpdateSubcommand < Cri::CommandRunner
```
Does it make sense to have a service update?
I don't have an idea on what would be useful for. For the moment I've left the skeleton of the subcommand here

miguelsorianod replied:
For the moment we are not going to implement this command unless we receive feedback that is needed with some use cases being used by users

eguzki commented on lib/3scale_toolbox/commands/copy_command/copy_service.rb:
```diff
@@ -29,7 +29,9 @@ def run
                                                  remote: threescale_client(source))
-          target_service = create_new_service(source_service.show_service, destination)
+          dest_service_params = source_service.show_service
+          dest_service_params["system_name"] = options[:target_system_name] if !options[:target_system_name].nil?
```
`dest_service_params["system_name"] = options[:target_system_name] unless options[:target_system_name].nil?`

eguzki commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,33 @@
+        def self.command
+          Cri::Command.define do
+            name        'create'
+            usage       'create'
```
something like this?
`create [opts] <source> <service-name>`

miguelsorianod replied:
:+1:

eguzki commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,33 @@
+          client = threescale_client(source)
+          result = client.create_service name: service_name
+          if !result['errors'].nil?
+            abort "Service has not been saved. Errors: #{result['errors']}"
```
`raise ThreeScaleToolbox::Error`. 

Let the toolbox exit the process and do some cleanup if it has to

eguzki replied:
why not use `Service.create`??

miguelsorianod replied:
:+1:

eguzki commented on lib/3scale_toolbox/commands/service_command/service_remove.rb:
```diff
@@ -0,0 +1,32 @@
+          service_id = arguments[:service_id]
+          client = threescale_client(source)
+
+          service = Entities::Service.new(id: service_id, remote: client)
```
find by system_name?? `service_found = Entities::Service::show_from_system_name(`

miguelsorianod replied:
:-1:

eguzki commented on lib/3scale_toolbox/commands/service_command/service_show.rb:
```diff
@@ -0,0 +1,99 @@
+          created_at updated_at
+        ]
+
+        def param_natural_number?(param)
```
There is `ThreeScaleToolbox::Helper` module for this kind of methods

miguelsorianod replied:
Finally we'll let the API raise a exception if incompatible

eguzki commented on lib/3scale_toolbox/commands/service_command/service_show.rb:
```diff
@@ -0,0 +1,99 @@
+            description "Show the information of a service"
+            runner ServiceShowSubcommand
+
+            param   :source
```
instead of `source` (which makes sense on copy command), what about `remote`???

miguelsorianod replied:
I chose `source` because a URI can be used instead of a remote name

eguzki replied:
`remote` for me it is either a remote name stored in toolbox config or URI. If it is confusing, we should state it clear in doc or cli help

miguelsorianod replied:
renamed to 'remote'

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_list.rb:
```diff
@@ -0,0 +1,32 @@
+        def self.command
+          Cri::Command.define do
+            name        'list'
+            usage       'list'
```
:+1:

eguzki commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,58 @@
+        def service_params
+          service_name = arguments[:service_name]
+          create_service_params = parse_options
+          create_service_params.merge!({name: service_name})
```
small detail, but, why not
`create_service_params[:name] = service_name`

miguelsorianod replied:
:+1:

eguzki commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,58 @@
+          remote = arguments[:remote]
+          client = threescale_client(remote)
+          create_service_params = service_params
+          result = client.create_service create_service_params
```
Double create???

miguelsorianod replied:
:+1:

eguzki commented on spec/unit/entities/service_spec.rb:
```diff
@@ -68,6 +70,73 @@
+    end
+  end
+
+  context 'Service.find_by_system_name' do
```
I think this is well tested :+1:

eguzki commented on spec/unit/entities/service_spec.rb:
```diff
@@ -68,6 +70,73 @@
+    let(:system_name) { random_lowercase_name }
+    let(:service_info) { { remote: remote, ref: system_name } }
+
+    it 'an exception is raised when remote is not configured' do
```
all of them tested :+1:

eguzki commented on spec/unit/entities/service_spec.rb:
```diff
@@ -68,6 +70,73 @@
+    let(:system_name) { random_lowercase_name }
+    let(:service_info) { { remote: remote, ref: system_name } }
+
+    it 'an exception is raised when remote is not configured' do
```
I would change the iteration text as: `remote call raises unexpected error`

miguelsorianod replied:
:+1:

eguzki commented on spec/unit/entities/service_spec.rb:
```diff
@@ -68,6 +70,73 @@
+
+    it 'service instance is returned when specifying an existing service ID' do
+      expect(remote).to receive(:show_service).and_return({"id" => system_name, "system_name" => "sysname1"})
+      expect(remote).not_to receive(:list_services)
```
without this negative expectation, the test would still fail when `list_services` are called, because remote is `double`

miguelsorianod replied:
:+1: I'll remove it

miguelsorianod commented on spec/unit/commands/service_command/list_command_spec.rb:
```diff
@@ -0,0 +1,30 @@
+require '3scale_toolbox'
+
+RSpec.describe ThreeScaleToolbox::Commands::ServiceCommand::ServiceListSubcommand do
+  include_context :random_name
```
:+1: removed

eguzki commented on spec/unit/commands/service_command/create_command_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+    # This should cover the case where the service already exists
+    # or other errors that are returned by calls to the API
+    context "when there is an error creating the service" do
```
AS commented above the test is correct for the current implementation. I would change implementation to handle `system_name has already been taken`.

This test would still be valid for any other kind of error

eguzki commented on spec/unit/commands/service_command/create_command_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      let(:service_create_args) { {"name" => service_name} }
+      let(:service_create_result) { {"name" => service_name, "system_name" => service_name, "id" => "1" } }
+
+      after :example do
```
Although still valid, `after` hook should be used to do cleanup kind of tasks.

If you want to share some expectation and DRY, the [shared_examples](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples) are the way to go.

miguelsorianod replied:
:+1: changed

miguelsorianod commented on lib/3scale_toolbox/commands/service_command/service_create.rb:
```diff
@@ -0,0 +1,51 @@
+          remote = arguments[:remote]
+          client = threescale_client(remote)
+          create_service_params = service_params
+          result = Entities::Service.create(remote: client, service_params: create_service_params)
```
I think making "create" idempotent is a little bit confusing.
I'm thinking of other tools that make "create" idempotent but I can't remember any of them.
For example, on kubernetes performing a "create" of already existing elements returns an error. It is not "idempotent". What they use instead is "patch" or "apply" which are equivalent to "update" and are completely different commands. the Kubernetes CLI tool must be used on lots of CI pipelines too.

Another option would be having a "-f" flag that when specified forces an "update" if the service already exists.

For the moment I'm leaving create "not idempotent".

eguzki replied:
:+1:

miguelsorianod commented:
@eguzki I've updated the PR with:
 * rebased changes on top of master
 * completely changed unit tests to mock the datamodel dependencies instead of the remote
 * added integration tests using assuming the real 3scale API is used only (not mocked versions)
 * Renamed some commands, file names and namespaces to follow the style that other CRUD commands created were following

Can you review it?

Documentation is still pending but I'll document it when the PR seems ok.

Thank you.

eguzki commented on spec/integration/commands/service_command/create_command_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+    after :example do
+      res = ThreeScaleToolbox::Entities::Service::find(remote: api3scale_client, ref: service_name)
+      res.delete
```
check `res` is not `nil`. It could happen if tests fails.

eguzki replied:
same for other `after` blocks

miguelsorianod replied:
Changed :+1:

eguzki commented on spec/integration/commands/service_command/create_command_spec.rb:
```diff
@@ -0,0 +1,67 @@
+    context 'without specifying options' do
+      let (:command_line_str) { "service create #{remote} #{service_name}" }
+
+      it "successfully creates a new service" do
```
I do not know if it makes sense. But, what about checking service options are expected default ones? WDYT?

eguzki commented:
Small nitpicks, but overall this is finished and approved. Great work :medal_military: !!

eguzki commented on lib/3scale_toolbox/commands/service_command/create_command.rb:
```diff
@@ -0,0 +1,59 @@
+            @remote ||= threescale_client(arguments[:remote])
+          end
+
+          def parse_options
```
I would merge `parse_options` and `service_attrs` like this (but it is ok now as it is)

```ruby
def service_attrs
  {
    "deployment_option" => options[:'deployment-mode'],
    "system_name" => options[:'system-name'],
    "backend_version" => options[:'authentication-mode'],
    "description" => options[:description],
    "support_email" => options[:'support-email'],
    "name" => arguments[:service_name],
  }.compact
end
```

miguelsorianod commented:
Added documentation for service CRUD

eguzki commented on docs/service.md:
```diff
@@ -0,0 +1,163 @@
+OPTIONS FOR SERVICE
+    -c --config-file=<value>              3scale toolbox configuration file
+                                          (default:
+                                          /home/msoriano/.3scalerc.yaml)
```
msoriano?

miguelsorianod replied:
fixed
